ast.stub.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+/**
+ * USE ast_stub.php INSTEAD IF YOU ARE LOOKING FOR DOCUMENTATION OR STUBS FOR YOUR IDE.
+ *
+ * This is a stub file meant only for use with https://github.com/php/php-src/blob/master/build/gen_stub.php
+ * to generate Reflection information (ReflectionParameter, ReflectionFunction, ReflectionMethod, etc.)
+ */
+
+// XXX: build/gen_stub.php does not support namespace nodes at the time of writing.
+// The namespace is thus deliberately omitted so that ast_arginfo.h can be successfully built.
+// namespace ast;
+
+
+/**
+ * Parses code file and returns AST root node.
+ *
+ * @param string $filename Code file to parse
+ * @param int    $version  AST version
+ * @return Node Root node of AST
+ *
+ * @see https://github.com/nikic/php-ast for version information
+ */
+function parse_code(string $code, int $version, string $filename = 'string code'): \ast\Node {}
+
+function parse_file(string $filename, int $version): \ast\Node {}
+
+/**
+ * @param int $kind AST_* constant value defining the kind of an AST node
+ * @return string String representation of AST kind value
+ */
+function get_kind_name(int $kind): string {}
+
+/**
+ * @param int $kind AST_* constant value defining the kind of an AST node
+ * @return bool Returns true if AST kind uses flags
+ */
+function kind_uses_flags(int $kind): bool {}
+
+/**
+ * Provides metadata for the AST kinds.
+ *
+ * The returned array is a map from AST kind to a Metadata object.
+ *
+ * @return Metadata[] Metadata about AST kinds
+ */
+function get_metadata(): array {}
+
+/**
+ * Returns currently supported AST versions.
+ *
+ * @param bool $exclude_deprecated Whether to exclude deprecated versions
+ * @return int[] Array of supported AST versions
+ */
+function get_supported_versions(bool $exclude_deprecated = false): array {}
+
+/**
+ * This class describes a single node in a PHP AST.
+ */
+class Node
+{
+    /**
+     * A constructor which accepts any types for the properties.
+     * For backwards compatibility reasons, all values are optional and can be any type, and properties default to null
+     *
+     * @param int|null $kind
+     * @param int|null $flags
+     * @param array|null $children
+     * @param int|null $lineno
+     */
+    public function __construct($kind = null, $flags = null, $children = null, $lineno = null) {
```

Comment:
Why no types on these?

Reply:
Mainly because this PR is intended to just add parameter defaults and reflect the types that are currently enforced at runtime

1. Adding types would be a backwards incompatible change not suitable for a patch version, for anything which passed unexpected types (e.g. "TODO: implement conversion for polyfill kind/flags") to the constructor.
2. There'd be a minor performance overhead to type checking or adding real types when calling the constructor that would apply to every generated node for a polyfill. I haven't benchmarked it.

This was also brought up in https://github.com/nikic/php-ast/pull/152#issuecomment-612510006

Reply:
We already do check types though, in https://github.com/nikic/php-ast/blob/f958360a2822a4f26fac5ddbb8ceecdcaea260fc/ast.c#L1245-L1251, so this doesn't make sense to me. We should be using the _THROW macro there though.

Reply:
Huh, I guess I forgot about that, so stubs should be safe. It throws a TypeError in php 8.0 and emits a warning (failing to initialize the node) in php 7.x.

I'm fine with throwing.